Feat(database-layer): create functions to perform crud operations for bot_types, bots, logs

kelpikz opened the pull request:
- This pr contains the layer for database layer in our application. [good article explaining our architecture](https://dzone.com/articles/layered-architecture-is-good)
- This contains various crud functions for the tables
   1. bot_types 
   2. bots
   3. logs
- From now on whenever we want to do any crud operation on our DB, we can simply call this function from our api-layer or gprc-layer.
- This is done to ensure that only one layer has access to the database, which helps us avoid any future bugs.
- Example workflow
   - Creating a new bot
      -   User makes a ```POST``` request to ```/api/bot/create``` with bot name and type
      -   The route handler for that route gets a db cursor from the connection pool and creates a new bot by calling ```create_bot``` in ```app/db/crud/bots.py```
      -   By doing this, the route handler logic and the persistence logic is separated, and the route handler only needs to worry about sending proper response to the user in the correct format.
 - RIght now the following crud functions are implemented along with validation
    1. Bot types
         1. creating bot type
         2. getting all bot types
         3. query bots with
             1. name
             2. id
         4. update bot type
         5. delete bot type
    2. bots
        1. create bot
        2. get all bots
        3. query bot with
             1. name
             2. bot type
        4. update bot data
        5. delete bot data
     3. logs
        1. create log
        2. fetch most recent logs with the following constraints
            1. log level
            2. bot
            3. limit

### To Test a method
  1. import ```Depends``` from ```fastapi```, ```get_connection``` from ```app/core/dependencies``` and relevant method you want to test
    
  ```py
  from fastapi import Depends
  from core.dependencies.database import get_connection
  from db.crud.bots import create_bot
  ```
  
  2. Call the function inside "/" route. Eg


  ```py
  @app.get("/")
  async def read_root(conn=Depends(get_connection)) -> Any:
      resp, err = await create_bot(conn, "test bot", 1)
      print(resp)
      print(err)
      return {"Dalal": "ToTheMoon"}
  ```

kelpikz committed:
feat(dependency): add pypika for query building

kelpikz committed:
fix(misc): rename variables and small bug fixes

kelpikz committed:
feat(crud): create add log crud function

kelpikz committed:
feat(core/dependency): create a FAST API dependency to obtain a connection from database connection pool

kelpikz committed:
fix(dependency): refactor get_connection dependency and fix few bugs

kelpikz committed:
feat(db): create a pydantic schema TimestampPlugin

kelpikz committed:
feat(db): create a model for getLogTailRequst to validate it

kelpikz committed:
feat(db): create helper function to convert db response to LogInDB model

kelpikz committed:
feat(db/log): create a function to query log tail

kelpikz committed:
misc: added some todods and helpful comments

kelpikz committed:
fix(db): refactor log crud functions

kelpikz committed:
fix(db): other refactor and quality of life changes for log crud functions

kelpikz committed:
feat(db): create a seperate file for useful errors

kelpikz committed:
feat(db): create functions to perfrom crud operations on bot_types table along with corresponding schemas

kelpikz committed:
misc(db): add some some todos for bot_types, which im too lazy to do now

kelpikz committed:
misc(db): add a todo in errors

kelpikz committed:
fix(db): minor bug fixes in crud/log.py

kelpikz committed:
feat(db): create a method to log sql query along with args

kelpikz committed:
feat(db): create functions to perform crud operations on bots along with corresponding schemas for data validation

kelpikz committed:
feat(lint): ingnore a flake error 'W503 line break before binary operator' in setup.cfg

kelpikz committed:
misc(lint): fix some lint error

kelpikz commented:
Need to add methods for updating and deleting bots and fix a few lint errors. In the meantime, please review all the other functions, as its quite extensive, and it probably takes a lot of time to review every function.

kelpikz committed:
feat(db): create method to update and delete bots

0149Sailesh commented on server/app/db/crud/bots.py:
```diff
@@ -0,0 +1,269 @@
+) -> Tuple[List[bots_schema.BotInDB], Union[Exception, None]]:
+    """Function to query uninflated details"""
+
+    logging.info(f"Quering bots with {name=} and {bot_type=} uninflated")
```
Remove the f before the string

kelpikz committed:
misc(lint): fix lint error

kelpikz committed:
misc(lint): fix lint error again

0149Sailesh commented on server/app/db/crud/bot_types.py:
```diff
@@ -0,0 +1,206 @@
+
+    logging.info(f"Trying to delete bot with {id=}")
+
+    q = MySQLQuery.from_(bot_types).delete().where(bot_types.id == id)
```
Shouldn't we clear all bots from bots table before deleting from bot_types?

kelpikz replied:
it will cascade delete

0149Sailesh replied:
Ahh right. But Ig we have to find a safe way to delete bots so that stocks don't get stranded in any of the bot accounts

kelpikz replied:
yea create an issue for it

0149Sailesh commented on server/app/db/crud/bot_types.py:
```diff
@@ -0,0 +1,206 @@
+        return None, e
+
+
+async def update_bot_name(
```
Change this to update_botType_name maybe? Thought it changes name in the bots table

0149Sailesh commented on server/app/db/crud/bots.py:
```diff
@@ -0,0 +1,331 @@
+        return [], None
+
+
+async def update_bot_name_with_id(
```
I think we also need a function to change the current type of bot to some other type of bot. For eg, Change a particular News Bot to ~Slave~ Bot

kelpikz replied:
lol okay

siva2204 commented on server/app/core/dependencies/database.py:
```diff
@@ -0,0 +1,31 @@
+    try:
+        yield conn
+    finally:
+        # BUG: Idk if we are supposed to close the connection after every request
```
idt we need to close db connection for every request.

does request and response has anything to do with db connections ?

kelpikz replied:
when we close a connection, it goes back to the pool and we can reuse it, this shouldn't be an issue

siva2204 replied:
okay then

siva2204 commented on server/app/db/crud/bot_types.py:
```diff
@@ -0,0 +1,206 @@
+bot_types = Table("bot_types")
+
+
+async def insert_bot_type(
```
adding it here for later reference, so we should register bot user in dalal server while inserting in db right ?

siva2204 commented on server/app/db/crud/bots.py:
```diff
@@ -0,0 +1,331 @@
+bots, bot_types = Table("bots", "bot_types")
+
+
+async def create_bot(
```
adding it here for later reference, so we should register bot user in dalal server while creating bot.

kelpikz replied:
do we need to add it here or should we add it some place else like ```api/controllers```?

siva2204 replied:
can we create something like this to handle all grpc unary calls

kelpikz replied:
yea we can do that in grpc manager

kelpikz committed:
feat(db/bots): update 'update_bot_name_with_id' method to update all the details of the bot, and add relevent schema for validating it

kelpikz committed:
fix(db): remove unwanted parameters in migrations,
fix: fix minor bug in settings
fix: DB_NAME value in .example.env
fix: fix start script in README.md

kelpikz committed:
docs(readme): add docs about env variables

kelpikz committed:
docs(readme): add app_env docs to server/readme

abesheknarayan commented on server/app/core/dependencies/database.py:
```diff
@@ -0,0 +1,31 @@
+        conn.close()
+
+
+async def get_connection(conn: Connection = Depends(_get_db_connection)) -> Cursor:
```
Do we have a need for cursors?

kelpikz replied:
That's the only way afasik to communicate to db when using connection pools according to [the docs](https://aiomysql.readthedocs.io/en/latest/pool.html#pool)

abesheknarayan commented on server/app/db/schemas/bot_types.py:
```diff
@@ -0,0 +1,34 @@
+
+
+def create_BotTypeInDB_from_tuple(data: Tuple[Any, ...]) -> BotTypeInDB:
+    """Utility function which creates a ButTypeInDB model from the database row."""
```
fix typo BotTypeInDB

kelpikz replied:
typo in the sense you want me to convert BotTypeInDB to snake case ?

abesheknarayan replied:
Utility function which creates a ~ButTypeInDB~ BotTypeInDB...

abesheknarayan commented:
@kelpikz write a clear description about what is added in this PR for future reference. Link any design docs if any.

abesheknarayan commented:
It would be better if we write unit / integration test parallely for every pr functionality. This would also help in creating a CI with the tests as now we dont have any means of knowing if things will be proper in the codebase when future changes come.

kelpikz commented on server/README.md:
```diff
@@ -0,0 +1,35 @@
+
+### Running Migrations
+
+- MySQL version **8.x.xx**
+- Make sure you have `PyMysqlClient` installed before running migrations
```
@siva2204 @0149Sailesh @abesheknarayan does this cover all the errors you guys got ?

kelpikz commented:
> It would be better if we write unit / integration test parallely for every pr functionality. This would also help in creating a CI with the tests as now we dont have any means of knowing if things will be proper in the codebase when future changes come.

Yes okay, will start working on it. Need to figure out how pytest works

abesheknarayan commented on server/app/db/schemas/bots.py:
```diff
@@ -0,0 +1,144 @@
+        is_valid_bot_type = False
+        if type(cls.bot_type is str):
+            if cls.bot_type == "":
+                is_valid_bot_type = False
```
change these if else's to single line? `is_valid_bot_type = (cls.bot_type == "")`

abesheknarayan commented on server/app/db/schemas/bots.py:
```diff
@@ -0,0 +1,144 @@
+            else:
+                is_valid_bot_type = True
+        else:
+            if cls.bot_type == 0:
```
likewise ^

kelpikz committed:
fix(db): fix import errors and minor bugs

kelpikz committed:
fix:remove unnecessary commit

kelpikz committed:
fix: fix minor bugs and add useful logs

kelpikz committed:
fix: fix minor bugs and add some extra log msgs, ran format and lint script

kelpikz committed:
feat: use logging.info instead of print to print useful data

kelpikz committed:
feat: use logging.info instead of print to print useful data

kelpikz committed:
fix: minor bug fixes and add extra documentation

kelpikz committed:
fix: lint fix

kelpikz committed:
feat(workflow): add mypy dependency installation step in workflow

kelpikz committed:
fix: update lint script to install mypy dependencies

kelpikz committed:
fix: update mypy missing import
